modoboa/core/api/v2/serializers.py

```diff
@@ -242,3 +279,201 @@ class UserAPITokenSerializer(serializers.Serializer):
+            email__iexact=clean_email, is_active=True)
+            .exclude(Q(secondary_email__isnull=True) | Q(secondary_email=""))
+        ).first()
+        if self.context["user"] is None:
+            raise NoUserFound()
+        return data
+
+    def save(self):
+        user = self.context["user"]
+        to_email = user.secondary_email
+        current_site = get_current_site(self.context["request"])
+        site_name = current_site.name
+        domain = current_site.domain
+        context = {
+            'email': to_email,
+            'domain': domain,
+            'site_name': site_name,
+            'uid': urlsafe_base64_encode(force_bytes(user.pk)),
+            'user': user,
+            'token': default_token_generator.make_token(user),
+            'protocol': 'https',
+        }
+        context.update(context_processors.new_admin_url())
+        subject = loader.render_to_string(
+            "registration/password_reset_subject.txt", context)
+        # Email subject *must not* contain newlines
+        subject = ''.join(subject.splitlines())
+        body = loader.render_to_string(
+            "registration/password_reset_email_v2.html", context)
+        try:
+            if send_mail(subject, body, None, [to_email]) == 0:
+                raise EmailFailedToSend()
+        except:
+            raise EmailFailedToSend()
+
+
+class SMSPasswordRecoveryInitSerializer(serializers.Serializer):
+
+    email = serializers.EmailField()
+
+    def validate(self, data, *args, **kwargs):
+        request = self.context["request"]
+
+        if not validate_email(data["email"]):
+            clean_email = data["email"]
+
+        self.context["user"] = (User.objects.filter(
+            email__iexact=clean_email, is_active=True)
+            .exclude(Q(phone_number__isnull=True) | Q(phone_number=""))
+        ).first()
+        if self.context["user"] is None:
+            raise NoSMSAvailible()
+
+        if not request.localconfig.parameters.get_value("sms_password_recovery"):
+            raise NoSMSAvailible()
+
+        user = self.context["user"]
+        if not user:
+            raise NoSMSAvailible()
+        backend = sms_backends.get_active_backend(
+            request.localconfig.parameters)
+        secret = cryptutils.random_hex_key(20)
+        code = oath.totp(secret)
+        text = _(
+            "Please use the following code to recover your Modoboa password: {}"
+            .format(code)
+        )
+        if not backend.send(text, [str(user.phone_number)]):
+            raise NoSMSAvailible()
+        request.session["totp_secret"] = secret
+        request.session["user_pk"] = user.pk
+        return data
+
+
+class PasswordRecoverySmsSerializer(serializers.Serializer):
+
+    sms_totp = serializers.CharField()
+
+    def validate(self, data):
+
+        if len(data['sms_totp']) != 6:
+            raise serializers.ValidationError(
+                "Wrong totp, try resend")
```

Comment:
I think you could achieve this by defining min_length and max_length options on the field definition itself